resources/channel.sub.js

```diff
@@ -0,0 +1,1003 @@
+            await this.sendMessage("postMessage", {msg: serialize(msg)}, false);
+        }
+
+        /**
+         * Disconnect the associated `RemoteGlobalCommandRecvChannel
+         * <#RemoteGlobalCommandRecvChannel>`_, if any, on the server
+         * side.
+         *
+         * @returns {Promise} - Resolved once the channel is disconnected.
+         */
+        disconnectReader() {
+            // This causes any readers to disconnect until they are explictly reconnected
+            return this.sendChannel.disconnectReader();
+        }
+
+        /**
+         * Close the channel and underlying websocket connections
+         */
+        close() {
+            let closers = [this.sendChannel.close()];
+            if (this.recvChannel !== null) {
+                closers.push(this.recvChannel.close());
+            }
+            if (this.respChannel !== null) {
+                closers.push(this.respChannel.close());
+            }
+            return Promise.all(closers);
+        }
+    }
+
+    self.RemoteGlobal = RemoteGlobal;
+
+    function typeName(obj) {
+        let type = typeof obj;
+        // The handling of cross-global objects here is broken
+        if (type === "undefined" ||
+            type === "string" ||
+            type === "boolean" ||
+            type === "number" ||
+            type === "bigint" ||
+            type === "symbol" ||
+            type === "function") {
+            return type;
+        }
+
+        if (obj === null) {
+            return "null";
+        }
+        if (obj instanceof RemoteObject) {
+            return "remoteobject";
+        }
+        if (obj instanceof SendChannel) {
+            return "sendchannel";
+        }
+        if (obj instanceof RecvChannel) {
+            return "recvchannel";
+        }
+        if (obj instanceof Error) {
+            return "error";
+        }
+        if (Array.isArray(obj)) {
+            return "array";
+        }
+        let constructor = obj.constructor && obj.constructor.name;
+        if (constructor === "RegExp" ||
+            constructor === "Date" ||
+            constructor === "Map" ||
+            constructor === "Set" ||
+            constructor == "WeakMap" ||
+            constructor == "WeakSet") {
+            return constructor.toLowerCase();
+        }
+        // The handling of cross-global objects here is broken
+        if (typeof window == "object" && window === self) {
+            if (obj instanceof Element) {
+                return "element";
+            }
+            if (obj instanceof Document) {
+                return "document";
+            }
+            if (obj instanceof Node) {
+                return "node";
+            }
+            if (obj instanceof Window) {
+                return "window";
+            }
+        }
+        if (Promise.resolve(obj) === obj) {
+            return "promise";
+        }
+        return "object";
+    }
+
+    let remoteObjectsById = new Map();
```

Comment:
We can't use `WeakMap` because we need to look by UUID, but won't this map keep objects alive? Can it be avoided?